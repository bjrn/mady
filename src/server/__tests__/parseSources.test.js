/* eslint-env jest */
import {
  _getRegexps as getRegexps,
  _parseFunctionCalls as parseFunctionCalls,
  _parseReactIntl as parseReactIntl,
} from '../parseSources';

describe('Source parser', () => {
  describe('Function parser', () => {
    it('should work in a simple case', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('Example');
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });

    it('should work with single and double quotes', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('Example');
        const str2 = FOO("Example2");
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });

    it('should extract context if available', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('context_Example');
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });

    it('should work with multiline strings', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('context_This is a very
long
string');
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });
  });

  describe('React Intl parser', () => {
    it('should extract messages from FormattedMessage components', () => {
      const keys = {};
      parseReactIntl(keys, 'FILENAME', `
        import { FormattedMessage } from 'react-intl';
        const name = 'Guille';
        const Component = () => (
          <FormattedMessage
            id="reactIntlId1"
            defaultMessage={\`context_Hello {NAME}!\`}
            values={{ NAME: <b>{name}</b> }}
            description="Bla bla bla"
          />
        );
      `);
      expect(keys).toMatchSnapshot();
    });

    it('should extract messages from FormattedHTMLMessage components', () => {
      const keys = {};
      parseReactIntl(keys, 'FILENAME', `
        import { FormattedHTMLMessage } from 'react-intl';
        const name = 'Guille';
        const Component = () => (
          <FormattedHTMLMessage
            id="reactIntlId2"
            defaultMessage={\`context_<i>Hello {NAME}!</i>\`}
            values={{ NAME: name }}
          />
        );
      `);
      expect(keys).toMatchSnapshot();
    });

    it('should extract messages from defineMessages() calls', () => {
      const keys = {};
      parseReactIntl(keys, 'FILENAME', `
        import { defineMessages } from 'react-intl';
        defineMessages({
          reactIntlId3: {
            id: 'reactIntlId3',
            description: 'Message to greet the user',
            defaultMessage: 'someContext_Hello, {NAME}!',
          },
        });
      `);
      expect(keys).toMatchSnapshot();
    });
  });
});
