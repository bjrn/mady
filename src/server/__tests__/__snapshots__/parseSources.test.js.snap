exports[`Source parser Function parser should extract context if available 1`] = `
Object {
  "Y29udGV4dF9FeGFtcGxl": Object {
    "context": "context",
    "firstUsed": null,
    "id": "Y29udGV4dF9FeGFtcGxl",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example",
    "unusedSince": null,
  },
}
`;

exports[`Source parser Function parser should work in a simple case 1`] = `
Object {
  "RXhhbXBsZQ==": Object {
    "context": null,
    "firstUsed": null,
    "id": "RXhhbXBsZQ==",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example",
    "unusedSince": null,
  },
}
`;

exports[`Source parser Function parser should work with multiline strings 1`] = `
Object {
  "Y29udGV4dF9UaGlzIGlzIGEgdmVyeQpsb25nCnN0cmluZw==": Object {
    "context": "context",
    "firstUsed": null,
    "id": "Y29udGV4dF9UaGlzIGlzIGEgdmVyeQpsb25nCnN0cmluZw==",
    "sources": Array [
      "FILENAME",
    ],
    "text": "This is a very
long
string",
    "unusedSince": null,
  },
}
`;

exports[`Source parser Function parser should work with single and double quotes 1`] = `
Object {
  "RXhhbXBsZQ==": Object {
    "context": null,
    "firstUsed": null,
    "id": "RXhhbXBsZQ==",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example",
    "unusedSince": null,
  },
  "RXhhbXBsZTI=": Object {
    "context": null,
    "firstUsed": null,
    "id": "RXhhbXBsZTI=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example2",
    "unusedSince": null,
  },
}
`;

exports[`Source parser React Intl parser should extract messages from FormattedHTMLMessage components 1`] = `
Object {
  "Y29udGV4dF88aT5IZWxsbyB7TkFNRX0hPC9pPg==": Object {
    "context": "context",
    "description": undefined,
    "firstUsed": null,
    "id": "Y29udGV4dF88aT5IZWxsbyB7TkFNRX0hPC9pPg==",
    "reactIntlId": "reactIntlId2",
    "sources": Array [
      "FILENAME",
    ],
    "text": "<i>Hello {NAME}!</i>",
    "unusedSince": null,
  },
}
`;

exports[`Source parser React Intl parser should extract messages from FormattedMessage components 1`] = `
Object {
  "Y29udGV4dF9IZWxsbyB7TkFNRX0h": Object {
    "context": "context",
    "description": "Bla bla bla",
    "firstUsed": null,
    "id": "Y29udGV4dF9IZWxsbyB7TkFNRX0h",
    "reactIntlId": "reactIntlId1",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Hello {NAME}!",
    "unusedSince": null,
  },
}
`;

exports[`Source parser React Intl parser should extract messages from defineMessages() calls 1`] = `
Object {
  "c29tZUNvbnRleHRfSGVsbG8sIHtOQU1FfSE=": Object {
    "context": "someContext",
    "description": "Message to greet the user",
    "firstUsed": null,
    "id": "c29tZUNvbnRleHRfSGVsbG8sIHtOQU1FfSE=",
    "reactIntlId": "reactIntlId3",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Hello, {NAME}!",
    "unusedSince": null,
  },
}
`;
